ext/console/02_console.js

```diff
@@ -805,24 +829,46 @@
     inspectOptions,
   ) {
     const gray = maybeColor(colors.gray, inspectOptions);
+    let lastValidIndex = 0;
+    let keys;
     const options = {
       typeName: "Array",
       displayName: "",
       delims: ["[", "]"],
-      entryHandler: (entry, inspectOptions, next) => {
+      entryHandler: (entry, inspectOptions) => {
         const [index, val] = entry;
         let i = index;
+        lastValidIndex = index;
         if (!ObjectPrototypeHasOwnProperty(value, i)) {
-          i++;
-          while (!ObjectPrototypeHasOwnProperty(value, i) && i < value.length) {
-            next();
-            i++;
+          let skipTo;
+          keys = keys || ObjectKeys(value);
+          if (keys.length === 0) {
+            // fast path, all items are empty
+            i = value.length;
+            skipTo = i;
+          } else {
+            // Not all indexes are empty or there's a non-index property
+            // Find first non-empty array index
+            i = value.length;
+            while (keys.length) {
+              const key = keys.shift();
+              // check if it's a valid array index
+              if (key > lastValidIndex && key < 2 ** 32 - 1) {
+                i = Number(key);
+                break;
+              }
+            }
```

Comment:
1) The spec allows up to (2⁵³ - 1), not sure if the engine does.
2) Can we agree that we absolutely don't want to parse numbers out of an array with a length such as.. (2¹⁵)? How about capping at the `iterableLimit` instead?

Reply:
1. For array indexes it's `2 ** 31 - 1`

https://262.ecma-international.org/6.0/#sec-object-type

> An array index is an integer index whose numeric value i is in the range +0 ≤ i < 2^32−1.

2. It would change the current output and it'll be different from Chrome's & Node's

Reply:
Ah. Well, on another note, I believe that the engine-provided integer indexes may be provided by `Array#keys`, so instead of `ObjectKeys` allocating an array, and iterating through until the right one is needed the iterator may be faster - also it gives integers, instead of strings, which may further help.

Reply:
`Object.keys` is exactly what we need here since it work for sparse arrays. It's the reason of the performance improvement.

```js
const x = new Array(5);
Object.keys(x); // [] empty
const keys = x.keys(); // Array#keys

keys.next(); // value: 0
keys.next(); // value: 1
// ... And so on.
```

Reply:
Right! My bad, I've since forgotten how `Array#keys` worked.